### PR TITLE
Fix and cleanup caching.

### PR DESCRIPTION
**Pull request by miekg**
Fix and cleanup caching. Cache differently for dnssec/non-dnssec and
UDP/TCP. Also add Fit function that trims a message to size (setting
TC when needed).

Drop the LRU caching a go with random eviction, eliminate all data races
and drop some of the useless cache metrics.

Add testing for the cache subpackage.

Cache stubforwarding, recursive and reverse requests.

Don't call EvictRandom under a lock This leads to deadlocks when the
cache is almost full. Drop the call to cache.Remove and just delete in
the map directly.

### CODE REVIEW

**Comment by miekg** [comment]
@kdima can you double check, please?

File: cache/cache.go [hunk 1]
```diff
@@ -21,65 +20,50 @@ import (
 // Elem hold an answer and additional section that returned from the cache.
 // The signature is put in answer, extra is empty there. This wastes some memory.
 type elem struct {
-	key        string
 	expiration time.Time // time added + TTL, after this the elem is invalid
 	msg        *dns.Msg
 }
 
 // Cache is a LRU cache that holds on the a number of RRs or DNS messgas.
```
Comment (kdima) [inline cache/cache.go]: it is not really lru right?

File: cache/cache.go [hunk 2]
```diff
@@ -21,65 +20,50 @@ import (
...
-		if e == nil { // nothing left
+// EvictRandom removes a random member a the cache.
+// Must be called under a write lock.
+func (c *Cache) EvictRandom() {
+	clen := len(c.m)
+	if clen < c.capacity {
+		return
+	}
+	i := c.capacity - clen
+	for k, _ := range c.m {
```
Comment (kdima) [inline cache/cache.go]: Check if you can delete no existing item

Comment (miekg) [inline cache/cache.go]: See <https://golang.org/ref/spec#Deletion_of_map_elements>, this will work.

File: cache/cache.go (hunk 1 above)
Comment (miekg) [inline cache/cache.go]: Nope. Fixed comments.

File: cache/cache.go [hunk 3]
```diff
@@ -89,14 +73,14 @@ func (c *Cache) InsertMessage(s string, msg *dns.Msg) {
 	if c.capacity == 0 {
```
Comment (kdima) [inline cache/cache.go]: capactiy can actually be negative I think

Comment (kdima) [inline cache/cache.go]: Same in some other places

File: cache/cache.go [hunk 4]
```diff
@@ -105,50 +89,50 @@ func (c *Cache) InsertSignature(s string, sig *dns.RRSIG) {
...
-	c.shrink()
+	c.EvictRandom()
+	c.Unlock()
 }
 
-// Search returns .... and a boolean indicating if we found something
+// Search returns a dns.Msg, the expiration time and a boolean indicating if we found something
 // in the cache.
 func (c *Cache) Search(s string) (*dns.Msg, time.Time, bool) {
 	if c.capacity == 0 {
```
Comment (kdima) [inline cache/cache.go]: This doesnt seem right. If the capacity is empty it means that we should definitely search

File: cache/cache.go [hunk 5]
```diff
@@ -105,50 +89,50 @@ func (c *Cache) InsertSignature(s string, sig *dns.RRSIG) {
...
 	}
-	c.Lock()
-	defer c.Unlock()
+	c.RLock()
 	if e, ok := c.m[s]; ok {
-		c.l.MoveToFront(e)
-		e := e.Value.(*elem)
-		e1 := e.msg.Copy()
-		return e1, e.expiration, true
+		c.RUnlock()
```
Comment (kdima) [inline cache/cache.go]: Maybe unlock shoud be deferred?

Comment (kdima) [inline cache/cache.go]: I think unlock before the copy can lead to a race condition. Copy then unlock

File: cache/cache.go [hunk 6]
```diff
@@ -105,50 +89,50 @@ func (c *Cache) InsertSignature(s string, sig *dns.RRSIG) {
...
 	}
+	if tcp {
+		i = append(i, byte(254))
+	}
 	return string(h.Sum(i))
 }
 
 // Key uses the name, type and rdata, which is serialized and then hashed as the key for the lookup.
-func Key(rrs []dns.RR) string {
+func KeyRRset(rrs []dns.RR) string {
```
Comment (kdima) [inline cache/cache.go]: Not used I think

File: cache/cache_test.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+		}
+		m1 = c.Hit(tc.m.Question[0], tc.dnssec, !tc.tcp, tc.m.Id)
+		if m1 != nil {
+			t.Fatalf("bad cache hit, expected <nil>, got %s:", m1)
+		}
+	}
+}
+
+func TestExpireMessage(t *testing.T) {
+	c := New(10, 1)
```
Comment (kdima) [inline cache/cache_test.go]: I owuld replace 1 with constant TTL. Then sleep wpuld be sleep(TTL+1)

File: cache/hit.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+	if hit {
+		// Cache hit! \o/
+		if time.Since(exp) < 0 {
+			m1.Id = msgid
+			m1.Compress = true
+			// Even if something ended up with the TC bit *in* the cache, set it to off
+			m1.Truncated = false
+			return m1
+		}
+		// Expired! /o\
```
Comment (kdima) [inline cache/hit.go]: +1

File: server/server.go [hunk 9]
```diff
@@ -212,77 +215,68 @@ func (s *server) ServeDNS(w dns.ResponseWriter, req *dns.Msg) {
...
+		metricSizeAndDuration(m1, start, tcp)
+		return
 	}
 
 	for zone, ns := range *s.config.stub {
 		if strings.HasSuffix(name, zone) {
 			resp := s.ServeDNSStubForward(w, req, ns)
-			metricSizeAndDuration(resp, start, tcp)
+			if resp != nil {
+				s.rcache.InsertMessage(cache.Key(q, dnssec, tcp), resp)
```
Comment (kdima) [inline server/server.go]: Key wrong arguments

**Comment by kdima** [comment]
LGTM

**Comment by miekg** [comment]
Thanks. Acted on basically all of your comments.

**Pull request merged**